### ISSUE TITLE
External custom controls

### ISSUE THREAD

**Issue by aschmidt93** [iljapostnovs/ui5plugin-parser#95]
Hi, 

I would like to use custom controls that are not part of the app, e.g. from NPM.

Consider the example repo https://github.com/SAP-samples/ui5-typescript-tutorial/tree/main/exercises/ex5/com.myorg.myapp.
There, the custom control (WindDirection) is defined as part of the app and the parser recognizes it without additional setup.
![image](https://github.com/iljapostnovs/ui5plugin-parser/assets/139967229/941334d2-3604-48c7-a03d-97ba86e9c184)

Now consider a custom control that is part of a UI5 TS library, e.g. https://github.com/SAP-samples/ui5-typescript-control-library. Let's assume the library is distributed via npm and is thus located at `./node_modules/ui5-typescript-control-library`.

Here are the example distributables: [ui5-typescript-control-library.zip](https://github.com/iljapostnovs/ui5plugin-parser/files/12098039/ui5-typescript-control-library.zip)

![image](https://github.com/iljapostnovs/ui5plugin-parser/assets/139967229/871f746c-4d99-41c6-8634-e17ea4661f91)

I have tried different values of the `additionalWorkspaces` property without sucess.

- `./node_modules/ui5-typescript-control-library` : Custom controls are not recognized - Error 
![image](https://github.com/iljapostnovs/ui5plugin-parser/assets/139967229/57fd406c-524e-4477-8329-0d22520a253a)

- `./node_modules/ui5-typescript-control-library/dist/resources` : Custom controls are recognized but attributes are not 
![image](https://github.com/iljapostnovs/ui5plugin-parser/assets/139967229/7250d01e-1ddf-4c6d-9542-7f0029ff9416)
- `./node_modules/ui5-typescript-control-library/dist/resources/com/myorg/myui5lib` : same as above

To reproduce I have set up a example repository: https://github.com/aschmidt93/ui5-example-cc

Note that the example lib is installed via file as it is not published to npm (see files above).

**Comment by iljapostnovs** [comment]
> 

Hi,

Thank you for providing very good example :)

This is a bit similar to https://github.com/iljapostnovs/ui5plugin-linter/issues/125, but still a bit different as well.
1) Is there a reason why this project is in `node_modules`?
2) In such format I see no way to make the parser work with this kind of structure. First of all, it is built project, not a project meant for development. 
It is TS project, but it doesn't have `tsconfig.json`. It is TS project, but it doesn't have types installed. It is TS project, but is has all kind of files mixed in: .d.ts, .ts, .js. In other words, it has a lot of missing and misleading information in order to parse it correctly. It is not recognized as TS project, because there is no `tsconfig.json`, so parser tries to parse it as JS project, and JS files are, well, like that:
![image](https://github.com/iljapostnovs/ui5plugin-parser/assets/31005247/fda3a894-8c57-49a8-91a8-c282d243e16e)

So yeah, plugin fails to recognize the structure there.

So with this example I could only suggest to download source code of the library as a separate project, (which has `tsconfig.json`, with installed TS types, and outside of the node_modules as a consiquence) and add it using `additionalWorkspaces`, e.g.
```json
"additionalWorkspaces": [
	"../ui5-typescript-control-library"
]
```

**Comment by aschmidt93** [comment]
Hello and thank you for your quick response. I have spent some time with the issue and debugged your plugin in order to better understand how it works. Here are my findings.

I have uploaded my fork of the SAP template for a UI5 Typescript library: https://github.com/aschmidt93/ui5-typescript-control-library

> Is there a reason why this project is in node_modules?

Yes, I'd like to have a UI5 Typescript library with custom controls that is distributed via NPM. The custom controls will be used by multiple projects. Thus, the library will be installed via npm and therefore be in the `node_modules` folder. The `ui5-tooling-modules` extension takes care of moving the files into the resources folder at runtime and during the build.

My assumption was that this is a fairly common use case to share reusable custom controls.

> First of all, it is built project, not a project meant for development.

Yes, but it does include .ts and .d.ts files (like a normal TS library). Hence it can be used for development, too.

> It is not recognized as TS project, because there is no tsconfig.json

I added the `tsconfig.json` to `node_modules/ui5-typescript-library` to resolve this issue. However, the `manifest.json` will not be found as of this line: https://github.com/iljapostnovs/ui5plugin-parser/blob/master/src/parser/abstraction/AbstractUI5Parser.ts#L90. During debugging I manually changed  `exclusions = ["**/node_modules/**]`. A parser was then created for the library. However, the attributes of the CC were still not parsed.

![image](https://github.com/iljapostnovs/ui5plugin-parser/assets/139967229/6fd56959-daad-4b56-9558-1ba214f17ed6)



> So with this example I could only suggest to download source code of the library as a separate project

I have changed the `package.json` of the library by removing the `files` property (which previously only included "dist" and "ui5.yaml like a normal TS library) , such that all files are now included when packaged.
Hence, the whole lib was now in the `node_modules`.

![image](https://github.com/iljapostnovs/ui5plugin-parser/assets/139967229/b57093a9-e402-4e6c-a7ff-7026008fe7f0)

However, the lib project was not recognized because no `manifest.json` exists. To resolve the issue I copied inside the lib  `./dist/resources/com/myorg/myui5lib/manifest.json` to `./src/com/myorg/myui5lib/manifest.json`. This resolved all issues and everything works. 

But this is a rather dirty hack.

IMO the parser should be able to parse a built UI5 Typescript library (like in the .zip file) as this is the "official" way to develop and distribute a UI5 Typesript library: https://github.com/SAP-samples/ui5-typescript-control-library.

**Comment by aschmidt93** [comment]
I see your problem, it is complicated indeed. Thank you for your effort.

> Is there any reason why you can't use source code instead?

 Usually, only the `dist` folder is published for a TS lib.

Until a better solution exists and as I love your extension, I will do it the following way:

- Change the `package.json` of the library by removing the `files` property, such that the whole source code is included
- Copy the `manifest.json` from `./dist/...` to root during the build process of the library
- Install the library via NPM
- Set in the project `"additionalWorkspaces": [
				"./node_modules/ui5-typescript-control-library"
			]`

**Comment by iljapostnovs** [comment]
> I see your problem, it is complicated indeed. Thank you for your effort.
> 
> > Is there any reason why you can't use source code instead?
> 
> Usually, only the `dist` folder is published for a TS lib. But I'm fairly new to creating UI5 TS libraries, so might be different there.
> 
> Until a better solution exists and as I love your extension, I will do it the following way:
> 
> * Change the `package.json` of the library by removing the `files` property, such that the whole source code is included
> * Copy the `manifest.json` from `./dist/...` to root during the build process of the library
> * Install the library via NPM
> * Set in the project `"additionalWorkspaces": [ "./node_modules/ui5-typescript-control-library" ]`

Usually only `dist` is published to `npm`, yes, but if it is open source project, usually node module comes with github repo, where you can find the source code. Maybe you have some real life public libraries which you would like to reuse? I could take a look at them.

**Comment by aschmidt93** [comment]
> Maybe you have some real life public libraries which you would like to reuse? I could take a look at them.

The library is public at npm but the source code is on the Gitlab server of the company I'm working at. The library was in plain JS so far and now we'd like to upgrade it to TS.

> I believe that there will be no issues in having source code for your own custom libraries, so this issue is only related to public open source libraries, am I correct? :)

You're right, there's no issue in that as we already had the JS files public on npm.

>From VSCode perspective it is just two folders:
One folder with library source code
Another with the app source code, and additionalWorkspaces references the library folder.
Works like a charm.

I see, so it is a monorepo. 

We (the company I'm working at) are  currently developing CAP apps https://cap.cloud.sap/docs/. Hence, we also use a monorepo with frontend and backend folders. However, the library we use is installed via npm as it is used by different projects.

**Comment by iljapostnovs** [comment]
> I see, so it is a monorepo.

No, it's not. It is other way around, as separated as possible :)
Two physical apps, two different Git repos, two different BSP Applications. 
But anyways, not sure how library topic works in CAP. Should work the same if Launchpad is used, but without launchpad I guess there will be no way to develop in another way without node modules or including the library physically into the project.

If it helps, I could create a separate parser for libs in node_modules, but as I mentioned previously it will only work if `manifest.json` is there and `-dbg.js` files are there.
The only issues I see there is that dependencies will not be parsed as expected, e.g. here:
![image](https://github.com/iljapostnovs/ui5plugin-parser/assets/31005247/d7c46ece-541b-48ee-9688-cd49642c20d8)
`Control` will work as expected, but `__ExampleRenderer, ___library` will not, because they are proxied with `_interopRequireDefault` function and the parser will not be able to distinguish what it means.

**Comment by iljapostnovs** [comment]
@aschmidt93, please check `nodeProjects` property. [Readme](https://github.com/iljapostnovs/ui5plugin-parser/blob/master/README.md#node-projects)
